### DISCUSSION TITLE
declarative constructor seems to behave inconsistently w/ collections such as attribute_mapped_collection when backrefs are used

### DISCUSSION THREAD

**Discussion by Estartu** [sqlalchemy/sqlalchemy#8369] [Usage Questions]
### Describe the bug

when I create an new object for an entity that is link via attribute_mapped_collection to another entity some times 
the object gets added with None as key. 

Every time I pass the foreign key object bevor the mapped attribute the new object gets added with None as a key. 

Month(month=1, year=year_object) works fine 
Month(year=year_object, month=1) adds the New object with None as a key. 

### To Reproduce

```python
from sqlalchemy import Column, create_engine, ForeignKey, Integer, SmallInteger
from sqlalchemy.orm import declarative_base, relation, sessionmaker
from sqlalchemy.orm.collections import attribute_mapped_collection

Base = declarative_base()
engine = create_engine("sqlite:///foo.db", echo=True)


class Month(Base):
    __tablename__ = 'months'
    id = Column(Integer, primary_key=True)
    year_id = Column(Integer, ForeignKey('years.id', ondelete="cascade"), nullable=False)
    year = relation('Year', back_populates='months')
    month = Column(SmallInteger, nullable=False)


class Year(Base):
    __tablename__ = 'years'
    id = Column(Integer, primary_key=True)
    year = Column(SmallInteger, nullable=False)
    months = relation('Month',
                      collection_class=attribute_mapped_collection('month'),
                      cascade="all, delete-orphan"
                      )


Base.metadata.create_all(engine)

Session = sessionmaker(bind=engine)
session = Session()

year = Year(year=2022)

print(year.months)

year.months[1] = Month(month=1)

print(year.months)

year.months[2] = Month(month=2, year=year)

print(year.months)

year.months[3] = Month(year=year, month=3)

session.commit()
```

### Error

```
/Users/estartu/PycharmProjects/WOS/Zeitkarten/bin/python /Users/estartu/Library/Application Support/JetBrains/PyCharm2022.2/scratches/scratch.py 
2022-08-09 14:40:38,763 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-08-09 14:40:38,764 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("months")
2022-08-09 14:40:38,764 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-08-09 14:40:38,764 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("years")
2022-08-09 14:40:38,765 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-08-09 14:40:38,765 INFO sqlalchemy.engine.Engine COMMIT
{}
{1: <__main__.Month object at 0x103176070>}
{1: <__main__.Month object at 0x103176070>, 2: <__main__.Month object at 0x103176ee0>}
{1: <__main__.Month object at 0x103176070>, 2: <__main__.Month object at 0x103176ee0>, None: <__main__.Month object at 0x1031764c0>, 3: <__main__.Month object at 0x1031764c0>}

Process finished with exit code 0
``` 

### Versions

- OS: MacOS 12.5
- Python: 3.9.13
- SQLAlchemy: 1.4.40
- Database: any
- DBAPI (eg: psycopg, cx_oracle, mysqlclient):

**Comment by zzzeek** [comment]
hey there -

In order to populate the attribute mapped collection, the "key" has to be available on the object; if not present, it will be None as you see.

The fact that "key" is present in the dictionary assignment is not actually used.  that is, if you did this:
     
    obj.collection[dictionary_key1] = RelatedObject(key=dictionary_key2)

Suppose dictionary_key1 and dictionary_key2 are different.  That's a conflict, what happens?  One of dictionary_key1 or dictionary_key2 has to win, right?   Or an error is thrown.   ~~As it turns out, dictionary_key2 wins here.~~ (edit: so it goes in the dictionary itself with dictionary_key1.  it's on a **backref** population that it goes in under dictionary_key2.  however, the acutal data for the DB no matter what is "dictionary_key2")   This is because that's how the object is loaded from the database also.     

Given that, down to your constructor, maybe you can see what might be happening.     [this is the default declarative construtor](https://github.com/sqlalchemy/sqlalchemy/blob/ac1d775d862e62561effcce54462f687dd79b8bb/lib/sqlalchemy/orm/decl_base.py#L1773).  there is not much to it and it assigns the given keyword arguments **in the order they were given**.  

So in your non-working case you basically have a rough equivalent of this:

```py
    m = Month()
    m.year = year  # backref populates object into year.months with key None
    m.month = 3
```

so a way to guard against this completely is to just add a helper constructor that makes sure month is set up, up front:

```py
class Month(Base):
    # . ..

    def __init__(self, **kw):
        self.month = kw.pop("month")
        super().__init__(**kw)
```

**Reply by Estartu** [reply]
shouldn't the back populate be delayed until all parameter are set in the constructor.

I see the problem in your example, but the order in the constructor shouldn't matter. 

This will cause hard to debug problems if someone uses ** with a prefilled dict to create database objects (which I doing quite often).

Month.month is defined as nullable=False. Therefor None shouldn't be considered a valid value and never be back populated. 

BTW the behavior change somewhere between 1.4.10 and now. I discovered the problem while updating a project to up to date versions. 1.4.10 didn't add the object at all.

Another Problem is that sqlalchemy tries do add the same object twice to the database,  resulting in a index error in my production case.

**Reply by zzzeek** [reply]
> shouldn't the back populate be delayed until all parameter are set in the constructor.

how would that work exactly?    The constructor is:

    self.foo = bar

Say you're just in your python code.  and you do this:

```py
m = Month()
m.year = year
print(year.months)
```

noting above, that's basically what a user-made constructor looks like, that is a use case that has to work:

```py
class Month:
    def __init__(self, year):
       self.year = year
```


What do you expect to see?  is there some extra step like this?

```py
m = Month()
m.year = year
sqlalchemy.run_all_back_populates(m)
print(year.months)
```

There's not really a way to make all the things happen here.    The best we could do, and maybe we should, is if someone did this:

```py

year.month[10] = Month(month=5)
```

it would throw an error that the given key and the effective key don't match.    That would be the best direction that can proceed here, if any.



> This will cause hard to debug problems if someone uses ** with a prefilled dict to create database objects (which I doing quite often).

OK, so I would advise dont use the default declarative constructor, use the manual one I gave you above.

> 
> Month.month is defined as nullable=False. Therefor None shouldn't be considered a valid value and never be back populated.

"None" was not populated anywhere significant here.  Add your Year to the database and flush, all the data is correct:

```py
>>> year.months[3] = m = Month(year=year, month=3)
>>> assert m.month == 3
>>> print(m.month)
3
>>> print(year.months)
{1: <__main__.Month object at 0x7f0befb89210>, 2: <__main__.Month object at 0x7f0befb88fa0>, None: <__main__.Month object at 0x7f0bf0f040d0>, 3: <__main__.Month object at 0x7f0bf0f040d0>}
>>> session.add(year)
>>> session.commit()
2022-08-09 11:28:28,357 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-08-09 11:28:28,358 INFO sqlalchemy.engine.Engine INSERT INTO years (year) VALUES (?)
2022-08-09 11:28:28,359 INFO sqlalchemy.engine.Engine [generated in 0.00011s] (2022,)
2022-08-09 11:28:28,362 INFO sqlalchemy.engine.Engine INSERT INTO months (year_id, month) VALUES (?, ?)
2022-08-09 11:28:28,362 INFO sqlalchemy.engine.Engine [generated in 0.00015s] (1, 1)
2022-08-09 11:28:28,362 INFO sqlalchemy.engine.Engine INSERT INTO months (year_id, month) VALUES (?, ?)
2022-08-09 11:28:28,362 INFO sqlalchemy.engine.Engine [cached since 0.0004889s ago] (1, 2)
2022-08-09 11:28:28,362 INFO sqlalchemy.engine.Engine INSERT INTO months (year_id, month) VALUES (?, ?)
2022-08-09 11:28:28,362 INFO sqlalchemy.engine.Engine [cached since 0.0007381s ago] (1, 3)
2022-08-09 11:28:28,363 INFO sqlalchemy.engine.Engine COMMIT
```

also you can assign None to any attribute in Python, SQLAlchemy can't assume that the business logic is not going to set the attribute to something non-None before the flush occurs, this is extremely common.

 
> BTW the behavior change somewhere between 1.4.10 and now. I discovered the problem while updating a project to up to date versions. 1.4.10 didn't add the object at all.

we can programmatically look for this, and I am not finding any 1.4 version that has different behavior, nor in main, nor in 1.3.  The behavior that is occurring here is admittedly not super intuitive but it's what's expected here for many years:

```py
year.months[3] = m = Month(year=year, month=3)

print(year.months)

# assert the behavior
assert year.months[None] is m


session.add(year)
session.commit()
```

now where you might see a difference is if you used **a different Python version**, like python 2 especially.  if you were using python 2 at some point, or possibly older python 3 versions like 3.5, which seems likely as I see you are using the very old `relation` symbol, then dictionaries / constructor behaviors are all going to be totally random.

**Reply by zzzeek** [reply]
note I edited "year" ,  i had the number 50 there for some reason.   which seems like neither a month or  a year

**Reply by zzzeek** [reply]
I just noticed that it's actually putting the same object in the dictionary twice, under both None and 3.  this is just what the dictionary view sees, data is still correct.  but this should have some kind of assertion.

**Reply by zzzeek** [reply]
that is, the backref sticks it under None, the dict assignment sticks it under "3".

**Reply by Estartu** [reply]
> "None" was not populated anywhere significant here. Add your Year to the database and flush, all the data is correct:

yes it is. Because i added the object manually to the mapped collection the object is twice in the collection and two inserts are generated by sqlalchemy. This produces either a duplicate row in the database or a index error as in my production system.

**Reply by zzzeek** [reply]
My test case against your code succeeds and there is no erroneous row, see [log output](https://github.com/sqlalchemy/sqlalchemy/discussions/8369#discussioncomment-3359194) above from my test, so there might be something else going on in your specific use case, would need to see it.

**Reply by zzzeek** [reply]
The reason it doesnt make sense that you'd have two INSERTs is that the unit of work system only inserts rows uniquely based on object identity, and it has nothing to do with if the collection has dupes in it, these are ignored unless the collection is a many-to-many.  demo:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)
    bs = relationship("B")


class B(Base):
    __tablename__ = 'b'
    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))
    data = Column(String)

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

b1 = B()
a1 = A()
a1.bs.append(b1)
a1.bs.append(b1)
a1.bs.append(b1)
a1.bs.append(b1)


assert len(a1.bs) == 4

s.add(a1)
s.commit()

assert len(a1.bs) == 1

```

**Reply by Estartu** [reply]
> The reason it doesnt make sense that you'd have two INSERTs is that the unit of work system only inserts rows uniquely based on object identity, and it has nothing to do with if the collection has dupes in it, these are ignored unless the collection is a many-to-many. demo:

when I wrote the code I'm updating right now, a newly created Month object wasn't added to the attribute_mapped_collection, therefor i used  
```python
amo=year.months.setdefault(month, Month(year=year, month=month)`
```

This code is in production right now and works fine with Python 3.8.9 and SQLAlchemy 1.4.10. I had a look at the git history. The code in production doesn't have the `back_populates='years'` in the relationship definition. That had been added after the build to fix sqlalchemy warnings.  

The Year object is loaded from the Database (Mariadb) and all Month are already there. If the code is executed in the collection are 13 different objects. One newly generated and 12 loaded from the database. Commiting the changes causes an index error becoase year_id and month are in an unique index. 

Even with the correct ordering of the arguments the code now doesn't work as expected, because it replaces the month object with the new one instead of discarding it. So the old entry in the database will be delated and an new one will be added. 

So am I getting it right to say that `back_populates='years'` causes the update of the MappedCollection when the relationship attribute is set. The old code works only because it's missing there.

I replaced the setdefault with

```python
amo = year.months.get(month) or Month(month=month, year=year)
```

That seams to work just fine.

**Reply by Estartu** [reply]
Another question just that I'm getting this strait 

```python 
Month(month=12, year=year)
```
updates the MappedCollection in year.months 

but 
```python 
year.month[7] = Month() 
```
is not expected to update the month or year attributes of the added object. 

Example code 
```python
 from sqlalchemy import Column, create_engine, ForeignKey, Integer, SmallInteger
from sqlalchemy.orm import declarative_base, relation, sessionmaker
from sqlalchemy.orm.collections import attribute_mapped_collection

Base = declarative_base()
engine = create_engine("sqlite:///foo.db", echo=True)


class Month(Base):
    __tablename__ = 'months'
    id = Column(Integer, primary_key=True)
    year_id = Column(Integer, ForeignKey('years.id', ondelete="cascade"), nullable=False)
    year = relation('Year', back_populates='months')
    month = Column(SmallInteger, nullable=False)


class Year(Base):
    __tablename__ = 'years'
    id = Column(Integer, primary_key=True)
    year = Column(SmallInteger, nullable=False)
    months = relation('Month',
                      collection_class=attribute_mapped_collection('month'),
                      cascade="all, delete-orphan"
                      )


Base.metadata.create_all(engine)

Session = sessionmaker(bind=engine)
session = Session()

year = Year(year=2022)

year.months[7] = Month()

print(year.months)

session.commit()

print(year.months[7].year, year.months[7].month)
```

Result 
```
2022-08-10 07:22:46,475 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2022-08-10 07:22:46,476 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("months")
2022-08-10 07:22:46,476 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-08-10 07:22:46,476 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("years")
2022-08-10 07:22:46,476 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-08-10 07:22:46,477 INFO sqlalchemy.engine.Engine COMMIT
{7: <__main__.Month object at 0x10d2991f0>}
None None
```

 I personaly whould expect either, but the later would be more likely exprected than the first

**Reply by zzzeek** [reply]
> 
> ```python
> amo=year.months.setdefault(month, Month(year=year, month=month)`
> ```
> 

right that's creating a Month object that you don't want, then w/ year the back_populates is pulling it in to your object graph which you don't want

> 
> ```python
> amo = year.months.get(month) or Month(month=month, year=year)
> ```

yes that's better.

**Reply by zzzeek** [reply]
> Another question just that I'm getting this strait
> 
> ```python
> Month(month=12, year=year)
> ```
> 
> updates the MappedCollection in year.months
> 
> but
> 
> ```python
> year.month[7] = Month() 
> ```
> 
> is not expected to update the month or year attributes of the added object.

that's correct.    having MappedCollection populate in both directions would be a feature add that is probably possible but we'd have to think about it.    I think MC is not super widely used so we don't get too many requests to improve it.    the main thing we'd have to do for this kind of use case, which we should if we had the motivation, is to augment the [event hook](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=attributeevents+append#sqlalchemy.orm.AttributeEvents.append) so that the key used in the dictionary assignment is passed along, so at least you could easily add this and we could later automate on the event hook.   but at the moment, the key used in `__getitem__()` isn't passed along, so you would need to subclass `MappedCollection` at the moment which is also doable but kind of tedious.

> I personaly whould expect either, but the later would be more likely exprected than the first

it should be more open ended at the very least, then the behavior could be an opt in

**Reply by zzzeek** [reply]
2.0 will have an event hook to make this possible.  as far as making it more of a parameter, that can be later but if you look at #8375 it has your case as a POC where the "month" is set using the event hook.

**Comment by zzzeek** [comment]
So there are two kinds of validation we can use to catch this.   One I think I'm going to add into 2.0 and make it the default.

The first is to validate that Month.month does not change if "year" is populated:

```py
from sqlalchemy.orm import validates

class Month(Base):
    __tablename__ = "months"
    id = Column(Integer, primary_key=True)
    year_id = Column(
        Integer, ForeignKey("years.id")
    )
    year = relationship("Year", back_populates="months")
    month = Column(SmallInteger, nullable=False)

    @validates("month")
    def _check_month(self, attrname, value):
        assert self.year is None, "Can't change month"
```

this is because attribute mapped dictionaries don't have any logic for intercepting changes to the key attribute in question.   so it makes sense this mutation should be prohibited.  if this were built-in somehow, there would be an option to turn it on.  not sure if there's a clean way to make it just an option without ripping everything open but will check.

the other is that instead of using `operator.attrgetter` for `Month.month`, we check if it's actually populated, as that None you see is not really there, it's just a default:

```py
from sqlalchemy import inspect
from sqlalchemy import exc
from sqlalchemy.orm.attri
def assertion_key_getter(key):
    def get_key(value):
        attr = inspect(value).attrs[key]
        attr_value = attr.loaded_value
        if attr_value is NO_VALUE:
            raise exc.InvalidRequestError(
                f"Key {key!r} for object {value} is not populated"
            )
        return value
    return get_key

class Year(Base):
    __tablename__ = "years"
    id = Column(Integer, primary_key=True)
    year = Column(SmallInteger, nullable=False)
    months = relationship(
        "Month",
        collection_class=lambda: MappedCollection(assertion_key_getter("month")),
        cascade="all, delete-orphan",
    )
```

that latter idea, I think I should just commit to 2.0 as the normal behavior as that's easy to add.

**Reply by Estartu** [reply]
you go way to much about the assignment in my example. The main problem is that 

`Month(year=year, month=3) `

produces a different result than 

`Month(month=3, year=year)`

and that is not expected in python. The order of the keywords shouldn't make a difference. 

the second problem is the None as key, which can't be happen if the objects are loaded from the Database. So it shouldn't be possible to produce it by simple supplying the keywords in the wrong order. 

I'm not that good in the internals of pythons, but keyword arguments are passed as a dictionary. The order in which the key are returned while iterating over the dict is based on the hash implementation of the dict and not the order they are given on the call.

I am not expecting sqlalchemy to update the Collection if i change the month nor would it be useful to prohibit changing the month. But i know no other place where the order of keyword arguments are significant to the result. That's not the python way. 

The constructor should assign all values given in the constructor and than execute any further processing of these attributes. This would make sure that the behavior is independent of the order in which the keywords are processed  (maybe given)

**Reply by zzzeek** [reply]
> you go way to much about the assignment in my example. The main problem is that
> 
> `Month(year=year, month=3) `
> 
> produces a different result than
> 
> `Month(month=3, year=year)`
> 
> and that is not expected in python. The order of the keywords shouldn't make a difference.

[pep-468](https://peps.python.org/pep-0468/), implemented since Python 3.6, states in the first paragraph (emphasis mine):

> The **kwargs syntax in a function definition indicates that the interpreter should collect all keyword arguments that do not correspond to other named parameters. However, Python does not preserved the order in which those collected keyword arguments were passed to the function. **In some contexts the order matters**. This PEP dictates that the collected keyword arguments be exposed in the function body as an ordered mapping.

Beyond that, there is no feasible way to reorganize the `__init__()` to defer backrefs without significant backwards incompatible changes that will significantly complicate the end-user experience of creating classes and mapping them, as well as add significant new codepaths and complexity to the ORM internals that would greatly increase maintenance and testing burden on our end.    Even if such a decision were made, it's not something that could happen for several years, so I would request that the particular issue of "constructor purity" be tabled.

I have a fix in review at #8372 that will fully prevent the issue you are seeing from occurring, which is intended to solve your actual problem.  your test above will fail very loudly in version 2.0 and additionally provide a parameter that will allow it to silently skip the erroneous backref assignment.

> 
> the second problem is the None as key, which can't be happen if the objects are loaded from the Database. So it shouldn't be possible to produce it by simple supplying the keywords in the wrong order.

with #8372 this particular flavor of None is completely prevented from being used.   has nothing to do with wrong order of keywords.

> 
> I'm not that good in the internals of pythons, but keyword arguments are passed as a dictionary. The order in which the key are returned while iterating over the dict is based on the hash implementation of the dict and not the order they are given on the call.

Again I would refer to https://peps.python.org/pep-0468/ where this ordering has been defined since Python 3.6 (as I mentioned in my [previous comment](https://github.com/sqlalchemy/sqlalchemy/discussions/8369#discussioncomment-3359194), this is a change in Python behavior since 3.6)


> 
> The constructor should assign all values given in the constructor and than execute any further processing of these attributes. This would make sure that the behavior is independent of the order in which the keywords are processed (maybe given)

as mentioned this would be a foundational rearchitecture of the ORM's assumptions that have been in place for decades and we aren't on board for doing somethign like that, nor is it really likely to be that helpful for the general issue as someone can assign to a backref outside of the constructor also.
